### ISSUE TITLE
Multithreading problem caused by FFTW dependency

### ISSUE THREAD

**Issue by VPetukhov** [JuliaStats/KernelDensity.jl#80]
**tl;dr** if you use many threads, running `FFTW.set_num_threads(1)` can be a good idea. Otherwise [FFTW](https://github.com/JuliaMath/FFTW.jl/) probably slows down computation and prevents outer parallelism. I suggest adding it to the README.

**Full explanation**
I was trying to do a lot of KDE in the loop, but it occurred that running the code in parallel slow down the process. Even if I simply set `JULIA_NUM_THREADS=20` (for 56 core server) without using `@threads`:

```julia
using KernelDensity
using Base.Threads

interp_kde(coords::Array{Float64, 2}, bandwidth::Float64) =
    InterpKDE(kde((coords[1,:], coords[2,:]), bandwidth=(bandwidth, bandwidth)))

td = rand(2, 100000);
@time for i in 1:500
    interp_kde(td, 1.0)
end
```

It creates multiple threads with loading 30% and takes 15.9 seconds. The same code with `JULIA_NUM_THREADS=1` takes 7.5 seconds, working fairly in single thread. Timing doesn't really change if I use `@threads:

```julia
@time @threads for i in 1:500
    interp_kde(td, 1.0)
end
```

After some digging, the problem occurred to be in the FFTW package, which is called somewhere during interpolation and by default [uses](https://github.com/JuliaMath/FFTW.jl/blob/20b5373d59cf4ce57442b3ac86258784700d6796/src/FFTW.jl#L59) `nthreads() * 4` threads inside its C code. To disable it you need to run `FFTW.set_num_threads(1)`. After that, running with  `JULIA_NUM_THREADS=20` but without `@threads` takes 7.5 seconds, as it should be, and with `@threads` it takes 0.5 seconds.

I was trying different run configurations, but at the end, looks like having FFTW parallel improves situation comparing to single thread only with large arrays (>500000) and large number of iterations (>100) And it's always much worse than having outer loop parallel.

**Comment by andreasnoack** [comment]
Which version of Julia are you using?

**Comment by VPetukhov** [comment]
1.3.1

**Comment by andreasnoack** [comment]
@stevengj Any thoughts on this?

**Comment by andreasnoack** [comment]
Related to https://github.com/JuliaStats/KernelDensity.jl/issues/123

**Comment by stevengj** [comment]
Recent versions of FFTW.jl [default to 1 thread](https://github.com/JuliaMath/FFTW.jl/blob/7b20cb548bae3abd3aa30278c2e22c4289e1d436/src/providers.jl#L62-L67).

(However, if you want to ensure that no one has changed this, you can pass `num_threads=1` to the `plan_fft` routines — this is better than setting the number of FFTW threads globally, which will affect other programs.)

Also, if you are repeatedly doing FFTs in a tight loop, you should really create the FFT plans once and re-use them.  Right now your API doesn't provide a way to do this if you are calling `kde(...)` over and over with different kernels of the same lengths, as in the example above — this would be a big speedup independent of threading.

**Comment by devmotion** [comment]
In the latest release we switched from FFTW to FFTA. On my computer the example above gives

```julia
julia> using KernelDensity

julia> using Base.Threads

julia> interp_kde(coords::Array{Float64, 2}, bandwidth::Float64) =
           InterpKDE(kde((coords[1,:], coords[2,:]), bandwidth=(bandwidth, bandwidth)))
interp_kde (generic function with 1 method)

julia> td = rand(2, 100000);

julia> @time for i in 1:500
           interp_kde(td, 1.0)
       end
  3.240257 seconds (5.40 M allocations: 3.045 GiB, 8.36% gc time, 98.95% compilation time)

julia> @time @threads for i in 1:500
           interp_kde(td, 1.0)
       end
  0.688493 seconds (527.77 k allocations: 2.807 GiB, 13.66% gc time, 31 lock conflicts, 45.55% compilation time)

julia> nthreads()
6
```
